src/lib/provable/crypto/encryption.ts

```diff
@@ -68,3 +73,95 @@ function decrypt(
 
   return message;
 }
+
+// v2
+
+function decryptV2(
+  { publicKey, cipherText }: CipherText,
+  privateKey: PrivateKey
+) {
+  // key exchange
+  let sharedSecret = publicKey.scale(privateKey.s);
+  let sponge = new Poseidon.Sponge();
+  sponge.absorb(sharedSecret.x);
+  let authenticationTag = cipherText.pop();
+
+  // decryption
+  let message = [];
+  for (let i = 0; i < cipherText.length; i++) {
+    let keyStream = sponge.squeeze();
+    let messageChunk = cipherText[i].sub(keyStream);
+
+    // convert to bytes
+    const byteMessage = wordToBytes(messageChunk, 32);
+
+    // pop frame bit
+    const frameBit = byteMessage.pop()!;
+
+    // check frame bit - if last element of the cipher text, frame bit must equal 1
+    // otherwise 0
+    if (i === cipherText.length - 1) frameBit.assertEquals(1);
+    else frameBit.assertEquals(0);
+    // push the message to our final message array
+    message.push(byteMessage);
+
+    if (i % 2 === 1) sponge.absorb(cipherText[i - 1]);
+    if (i % 2 === 1 || i === cipherText.length - 1)
+      sponge.absorb(cipherText[i]);
+  }
+
+  // authentication tag
+  sponge.squeeze().assertEquals(authenticationTag!);
+
+  // return the message as a flat array of bytes
+  return message.flat();
```

Comment:
hm, we're not stripping off the padding here. it will make the function a bit hard to deal with. but I think it's the right approach, the actual message length needs to be hard-coded in the circuit anyway

Reply:
yep, still need to do that. I am thinking of adding a property to `CipherText`